compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -285,19 +285,36 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
             TyKind::MacCall(..) => self.visit_macro_invoc(ty.id),
             TyKind::ImplTrait(node_id, _) => {
                 let parent_def = match self.impl_trait_context {
-                    ImplTraitContext::Universal(item_def) => self.resolver.create_def(
-                        item_def,
-                        node_id,
-                        DefPathData::ImplTrait,
-                        self.expansion.to_expn_id(),
-                        ty.span,
-                    ),
+                    ImplTraitContext::Universal(item_def) => {
+                        //| ImplTraitContext::UniversalInDyn(item_def) => {
+                        let def_id = self.resolver.create_def(
+                            item_def,
+                            node_id,
+                            DefPathData::ImplTrait,
+                            self.expansion.to_expn_id(),
+                            ty.span,
+                        );
+                        self.resolver
+                            .impl_trait_context
+                            .insert(def_id, ImplTraitContext::Universal(item_def));
+                        def_id
+                    }
                     ImplTraitContext::Existential => {
-                        self.create_def(node_id, DefPathData::ImplTrait, ty.span)
+                        let def_id = self.create_def(node_id, DefPathData::ImplTrait, ty.span);
+                        self.resolver
+                            .impl_trait_context
+                            .insert(def_id, ImplTraitContext::Existential);
+                        def_id
                     }
+                    ImplTraitContext::UniversalInDyn(_) => self.parent_def,
                 };
                 self.with_parent(parent_def, |this| visit::walk_ty(this, ty))
             }
+            TyKind::TraitObject(ref bounds, ..) => {
+                self.with_impl_trait(ImplTraitContext::UniversalInDyn(self.parent_def), |this| {
+                    walk_list!(this, visit_param_bound, bounds, BoundKind::TraitObject);
```

Comment:
```suggestion
                    visit::walk_ty(self, ty)
```
No need to re-code the default behaviour.